domain-models-runtime/src/main/java/org/folio/rest/impl/TenantAPI.java

```diff
@@ -319,14 +321,23 @@ public void postTenant(TenantAttributes tenantAttributes, Map<String, String> he
               ? PostTenantResponse.respond200WithApplicationJson(jsonListOfFailures)
               : PostTenantResponse.respond201WithApplicationJson(jsonListOfFailures);
     })
+
     .onFailure(e -> {
       if (e instanceof NoSchemaJsonException) {
         handlers.handle(Future.succeededFuture(PostTenantResponse.respond204()));
         return;
       }
       log.error(e.getMessage(), e);
-      handlers.handle(Future.succeededFuture(PostTenantResponse.respond500WithTextPlain(e.getMessage())));
+      String text = e.getMessage() + "\n" + ExceptionUtils.getStackTrace(e);
+      Response response = PostTenantResponse.respond500WithTextPlain(text);
+      handlers.handle(Future.failedFuture(new ResponseException(response)));
     })
-    .onSuccess(response -> handlers.handle(Future.succeededFuture(response)));
+    .onSuccess(response -> {
+      if (response.getStatus() >= 300) {
+        handlers.handle(Future.failedFuture(new ResponseException(response)));
+        return;
+      }
+      handlers.handle(Future.succeededFuture(response));
+    });
```

Comment:
Does this mean that all clients will need to change their failure handling?

Reply:
No, clients that have a failure can pass the 400-/500-Response either as a succeeding AsyncResult or wrapped in ResponseException as a failing AsyncResult. RestVerticle#getResponse extracts it from all possibilities.